File: rest-server/src/main/java/org/cifpaviles/pokedam/rest_server/game/service/LobbyService.java
package org.cifpaviles.pokedam.rest_server.game.service;

import java.util.Collection;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

import org.cifpaviles.pokedam.rest_server.models.Lobby;
import org.cifpaviles.pokedam.rest_server.models.LobbyInfo;
import org.cifpaviles.pokedam.rest_server.models.LobbyJoined;
import org.cifpaviles.pokedam.rest_server.models.LobbyLeft;
import org.cifpaviles.pokedam.rest_server.models.Joiner;
import org.cifpaviles.pokedam.rest_server.entity.User;
import org.cifpaviles.pokedam.rest_server.repository.UserRepository;
import org.springframework.stereotype.Service;

@Service
public class LobbyService {
    // Almacena todas las partidas por ID
    private final Map<String, Lobby> lobbies = new ConcurrentHashMap<>();

    // Rastrea a qué partida pertenece cada jugador
    private final Map<Long, String> playerRooms = new ConcurrentHashMap<>();

    private final UserRepository userRepository;

    public LobbyService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    private String getPlayerNickname(Long playerId) {
        return userRepository.findById(playerId)
                .map(User::getNickname)
                .orElse("????");
    }

    public Map.Entry<String, Lobby> create(Long playerId, String name, String password) {
        // Si ya está en una sala antes, lo sacamos

        leave(playerId);

        String roomId = UUID.randomUUID().toString();

        Lobby lobby = new Lobby(name, password, playerId, getPlayerNickname(playerId));

        lobbies.put(roomId, lobby);
        playerRooms.put(playerId, roomId);
        return Map.entry(roomId, lobby);
    }

    public LobbyJoined join(String lobbyId, Long playerId, String passwordInput) {
        Lobby lobby = lobbies.get(lobbyId);

        if (lobby != null) {

            // Verificación simple de contraseña. Si hay y no coincide, falla.
            if (lobby.password != null && !lobby.password.equals(passwordInput)) {
                return null;
            }
            var player = new Joiner(false, getPlayerNickname(playerId));

            lobby.joiners.put(playerId, player);
            playerRooms.put(playerId, lobbyId);
            return new LobbyJoined(lobby, player);
        }
        return null;
    }

    public Map.Entry<String, Lobby> setReady(Long playerId, boolean isReady) {
        String roomId = playerRooms.get(playerId);
        if (roomId != null) {
            Lobby room = lobbies.get(roomId);
            if (room != null) {
                var player = room.joiners.get(playerId);
                if (player != null) {
                    player.isReady = isReady;
                }
                return Map.entry(roomId, room);
            }
        }
        return null;
    }

    public Map.Entry<String, Lobby> startGame(Long playerId) {
        String roomId = playerRooms.get(playerId);
        if (roomId != null) {
            Lobby room = lobbies.get(roomId);

            // Revisa que sea el host y esté WAITING
            if (room != null && room.hostId.equals(playerId)) {
                // Validación para iniciar: ejemplo, al menos 2 en partida, y todos readys =
                // true
                Collection<Joiner> allReadyStates = room.joiners.values();
                if (allReadyStates.size() > 1 && allReadyStates.stream().allMatch(player -> player.isReady)) {
                    return Map.entry(roomId, room);
                }
            }
        }
        return null;
    }

    public LobbyLeft leave(Long playerId) {
        String roomId = playerRooms.remove(playerId);
        if (roomId == null)
            return null;

        Lobby room = lobbies.get(roomId);
        if (room.hostId == playerId) {
            if (room.joiners.isEmpty()) {
                lobbies.remove(roomId);
                return new LobbyLeft(roomId, null);
            } else {
                Map.Entry<Long, Joiner> newHostEntry = room.joiners.entrySet().iterator().next();
                room.hostId = newHostEntry.getKey();
                room.hostNickname = newHostEntry.getValue().nickname;
                room.joiners.remove(newHostEntry.getKey());
            }
            return new LobbyLeft(roomId, room, true);
        } else {
            if (room.joiners.remove(playerId) == null)
                return null;

            return new LobbyLeft(roomId, room);
        }
    }

    public Collection<LobbyInfo> getAll() {
        return lobbies.entrySet().stream().map(entry -> {
            String roomId = entry.getKey();
            Lobby lobby = entry.getValue();
            return new LobbyInfo(roomId, lobby);
        }).collect(Collectors.toList());
    }

    public Lobby get(String roomId) {
        return lobbies.get(roomId);
    }
}

File: backend/src/services/LobbyService.js
const { v4: uuidv4 } = require('uuid');
const { User } = require('../config/Database');

class LobbyService {
    constructor() {
        this.lobbies = new Map();
        this.playerRooms = new Map();
    }

    async getPlayerNickname(playerId) {
        const user = await User.findByPk(playerId) ?? await User.create({ 
            nickname: `Trainer${String(playerId).padStart(4, '0')}`,
            refreshToken: uuidv4(),
        });
        return user.nickname;
    }

    async create(playerId, name, password) {
        // Si ya está en una sala antes, lo sacamos
        this.leave(playerId);

        const roomId = uuidv4();
        // Al crear ya deberíamos tener el name desde el UserDB para no hacer query síncrona aquí si se puede evitar
        const nickname = await this.getPlayerNickname(playerId);

        const lobby = {
            name,
            password: password || null,
            hostId: playerId,
            hostNickname: nickname,
            state: 'WAITING',
            joiners: new Map(), // playerId -> { isReady, nickname }
            maxPlayers: 8
        };

        this.lobbies.set(roomId, lobby);
        this.playerRooms.set(playerId, roomId);

        return { key: roomId, value: lobby };
    }

    async join(lobbyId, playerId, passwordInput) {
        const lobby = this.lobbies.get(lobbyId);

        if (lobby) {
            if (lobby.password && lobby.password !== passwordInput) {
                return null;
            }

            const nickname = await this.getPlayerNickname(playerId);

            const player = {
                isReady: false,
                nickname: nickname
            };

            lobby.joiners.set(playerId, player);
            this.playerRooms.set(playerId, lobbyId);

            return { lobby, joiner: player };
        }
        return null;
    }

    setReady(playerId, isReady) {
        const roomId = this.playerRooms.get(playerId);
        if (roomId) {
            const room = this.lobbies.get(roomId);
            if (room) {
                const player = room.joiners.get(playerId);
                if (player) {
                    player.isReady = isReady;
                }
                return { key: roomId, value: room };
            }
        }
        return null;
    }

    startGame(playerId) {
        const roomId = this.playerRooms.get(playerId);
        if (roomId) {
            const room = this.lobbies.get(roomId);

            // Revisa que sea el host
            if (room && room.hostId === playerId) {
                const allReadyStates = Array.from(room.joiners.values());
                // Ejemplo: al menos 1 joiner (2 en total) y todos ready
                if (allReadyStates.length > 0 && allReadyStates.every(player => player.isReady)) {
                    room.state = 'IN_GAME';
                    return { key: roomId, value: room };
                }
            }
        }
        return null;
    }

    leave(playerId) {
        const roomId = this.playerRooms.get(playerId);
        if (!roomId) return null;

        this.playerRooms.delete(playerId);
        const room = this.lobbies.get(roomId);

        if (room.hostId === playerId) {
            if (room.joiners.size === 0) {
                this.lobbies.delete(roomId);
                return { lobbyId: roomId, lobby: null };
            } else {
                // Nuevo host al azar del primer elemento
                const firstEntry = room.joiners.entries().next().value;
                const newHostId = firstEntry[0];
                const newHostData = firstEntry[1];

                room.hostId = newHostId;
                room.hostNickname = newHostData.nickname;
                room.joiners.delete(newHostId);

                return { lobbyId: roomId, lobby: room, hostReplacement: true };
            }
        } else {
            const removed = room.joiners.delete(playerId);
            if (!removed) return null;

            return { lobbyId: roomId, lobby: room };
        }
    }

    getAll() {
        const all = [];
        for (const [id, lobby] of this.lobbies.entries()) {
            all.push({
                id,
                name: lobby.name,
                hasPassword: !!lobby.password,
                playerCount: lobby.joiners.size + 1,
                maxPlayers: lobby.maxPlayers
            });
        }
        return all;
    }

    get(roomId) {
        const lobby = this.lobbies.get(roomId);
        if (!lobby) return null;

        // Convertir el Map de joiners a object literal para el JSON
        const joinersObj = {};
        for (const [pid, player] of lobby.joiners.entries()) {
            joinersObj[pid] = player;
        }

        return {
            name: lobby.name,
            state: lobby.state,
            hostId: lobby.hostId,
            hostNickname: lobby.hostNickname,
            maxPlayers: lobby.maxPlayers,
            joiners: joinersObj
        };
    }
}

// Exportamos una única instancia (Singleton pattern, como el @Service de Spring)
module.exports = new LobbyService();
